### DISCUSSION TITLE
Callable[Mapping[str, Any], None] is incompatible with func(a: TypedDict).  Is this a bug?

### DISCUSSION THREAD

**Discussion by deuscapturus** [microsoft/pyright#5514] [Q&A] [answered]
[PEP589](https://peps.python.org/pep-0589/#type-consistency) reads:
> First, any TypedDict type is consistent with Mapping[str, object].

This isn't true when comparing function signatures.  Is this a bug or a limitation of Callable?

```python
#!/usr/bin/python3.11

from typing import TypedDict, Mapping, Callable, Any

class A(TypedDict):
    config: str

def a_func(a: A) -> None:
    pass

def a_callback(a: Callable[[Mapping[str, object]], None]):
    pass

a_callback(a_func)
# 👆 Error:  type "Mapping[str, object]" cannot be assigned to type "A". "Mapping[str, object]" is incompatible with "A"
```

**Comment by erictraut** [comment, marked as answer]
Parameters in a callback function are contravariant, not covariant. That means the parameter type for the supplied function must be a supertype, not a subtype, of the declared callback type.

In your example, the function `a_callback` requires you to pass a function that accepts any `Mapping[str, object]`, but you're instead passing it a function that accepts only type `A` (or a subtype thereof). If `a_callback` were to call your callback and pass it some other `Mapping[str, object]` that is not `A`, it would violate the signature of `a_func`.
